src/core/layertreemodel.cpp

```diff
@@ -691,6 +692,26 @@ QVariant FlatLayerTreeModelBase::data( const QModelIndex &index, int role ) cons
       if ( !layer ) // Group
         return -1;
 
+      // For now, do not count feature on WFS layers, it can lead to long hangs
+      if ( layer->dataProvider() && layer->dataProvider()->name() == QStringLiteral( "WFS" ) )
+        return QVariant();
+
+      if ( layer->renderer() && layer->renderer()->legendSymbolItems().size() > 0 )
+      {
+        long count = layer->featureCount( layer->renderer()->legendSymbolItems().at( 0 ).ruleKey() );
+        if (  count == -1 )
+        {
+          layer->countSymbolFeatures();
+
+          const QTime dieTime = QTime::currentTime().addMSecs(300);
+          while ( QTime::currentTime() < dieTime && count == -1 )
+          {
+            QCoreApplication::processEvents(QEventLoop::AllEvents, 50);
```

Comment:
https://doc.qt.io/qt-5/qtquick-performance.html#timing-considerations

> Note: A pattern which is tempting, but should never be used, is creating your own QEventLoop or calling QCoreApplication::processEvents() in order to avoid blocking within a C++ code block invoked from QML. This is dangerous, because when an event loop is entered in a signal handler or binding, the QML engine continues to run other bindings, animations, transitions, etc. Those bindings can then cause side effects which, for example, destroy the hierarchy containing your event loop.

Reply:
@m-kuhn , any suggestion on an alternative approach to provide a small wait period here? Basically, without it, the first time someone opens the layer properties panel, he/she will not see any feature count.

Reply:
I guess we can add a timer _in the QML itself_.

Reply:
Use the `symbolFeatureCountMapChanged` signal to get a notification when the new count arrives. Emit a `dataChanged` signal from the model (or add a new property, whatever is more suitable).